### ISSUE TITLE
[MNG-6216] ArrayIndexOutOfBoundsException when parsing POM 

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7233]
**[Marc Savy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msavy)** opened **[MNG-6216](https://issues.apache.org/jira/browse/MNG-6216?redirect=false)** and commented

When building apiman https://github.com/apiman/apiman.git on Maven 3.5.0 the XML parser seems to have some issues parsing the apiman-test-common module.

If you want to reproduce, you can simply use mvn clean install, no profiles needed.

Log: https://gist.github.com/msavy/a1b4307fb238a543dfd6af426b42d446

Also attached.

---

**Affects:** 3.5.0

**Attachments:**
- [log](https://issues.apache.org/jira/secure/attachment/12862793/log) (_42.06 kB_)

**Sub-tasks:**
- [MNG-6280](https://issues.apache.org/jira/browse/MNG-6280) ArrayIndexOutOfBoundsException caused by pom.xml with process instructions

- [MNG-6281](https://issues.apache.org/jira/browse/MNG-6281) ArrayIndexOutOfBoundsException caused by pom.xml with invalid/duplicate XML

**Issue Links:**
- [MNG-2802](https://issues.apache.org/jira/browse/MNG-2802) Concurrent-safe access to local Maven repository
 (_**"is fixed by"**_)
- [MNG-6243](https://issues.apache.org/jira/browse/MNG-6243) Build fails with ArrayIndexOutOfBoundsException: 53090
 (_**"is duplicated by"**_)
- [MNG-6338](https://issues.apache.org/jira/browse/MNG-6338) Improve error message
 (_**"is duplicated by"**_)
- [MNG-6281](https://issues.apache.org/jira/browse/MNG-6281) ArrayIndexOutOfBoundsException caused by pom.xml with invalid/duplicate XML

- [MNG-6176](https://issues.apache.org/jira/browse/MNG-6176) Javadoc errors prevent release with Java 8
 (_**"is superceded by"**_)

**Remote Links:**
- [Plexus Utils #22
](https://github.com/codehaus-plexus/plexus-utils/issues/22)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

0 votes, 11 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I cannot reproduce it on Windows. Please provide more input.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hi, I can not reproduce this as well (on Mac OS)...getting different failures at while not finding some artifacts. The mentioned module which failed at your site is building fine here...
(Apart from some comments I would make this build which is a different story)...

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] apiman ............................................. SUCCESS [  1.634 s]
[INFO] apiman-common ...................................... SUCCESS [  0.012 s]
[INFO] apiman-common-auth ................................. SUCCESS [  4.142 s]
[INFO] apiman-common-util ................................. SUCCESS [  2.203 s]
[INFO] apiman-common-config ............................... SUCCESS [  6.966 s]
[INFO] apiman-common-net .................................. SUCCESS [  7.000 s]
[INFO] apiman-common-plugin ............................... SUCCESS [  1.866 s]
[INFO] apiman-common-servlet .............................. SUCCESS [  1.507 s]
[INFO] apiman-common-logging .............................. SUCCESS [  0.006 s]
[INFO] apiman-common-logging-core ......................... SUCCESS [  1.560 s]
[INFO] apiman-gateway ..................................... SUCCESS [  0.006 s]
[INFO] apiman-gateway-engine .............................. SUCCESS [  0.005 s]
[INFO] apiman-gateway-engine-beans ........................ SUCCESS [  7.237 s]
[INFO] apiman-gateway-engine-core ......................... SUCCESS [  9.698 s]
[INFO] apiman-common-logging-slf4j ........................ SUCCESS [  0.967 s]
[INFO] apiman-common-logging-log4j2 ....................... SUCCESS [  1.152 s]
[INFO] apiman-distro ...................................... SUCCESS [  0.005 s]
[INFO] apiman-distro-data ................................. SUCCESS [  0.124 s]
[INFO] apiman-manager ..................................... SUCCESS [  0.007 s]
[INFO] apiman-manager-api ................................. SUCCESS [  0.008 s]
[INFO] apiman-manager-api-beans ........................... SUCCESS [ 18.517 s]
[INFO] apiman-distro-ddl .................................. SUCCESS [ 20.540 s]
[INFO] apiman-manager-api-config .......................... SUCCESS [  1.507 s]
[INFO] apiman-manager-api-core ............................ SUCCESS [  3.823 s]
[INFO] apiman-gateway-api ................................. SUCCESS [  0.004 s]
[INFO] apiman-gateway-api-rest ............................ SUCCESS [  1.623 s]
[INFO] apiman-manager-api-gateway ......................... SUCCESS [  2.014 s]
[INFO] apiman-manager-api-export-import-api ............... SUCCESS [  2.015 s]
[INFO] apiman-manager-api-security ........................ SUCCESS [  2.338 s]
[INFO] apiman-manager-api-rest ............................ SUCCESS [  2.541 s]
[INFO] apiman-test ........................................ SUCCESS [  0.005 s]
[INFO] apiman-test-common ................................. SUCCESS [  5.587 s]
[INFO] apiman-manager-api-migrator ........................ SUCCESS [  2.466 s]
[INFO] apiman-manager-api-rest-impl ....................... SUCCESS [  3.600 s]
[INFO] apiman-manager-api-jpa ............................. SUCCESS [  5.556 s]
[INFO] apiman-manager-api-es .............................. SUCCESS [  7.034 s]
[INFO] apiman-manager-api-jdbc ............................ SUCCESS [  2.115 s]
[INFO] apiman-manager-api-hawkular ........................ SUCCESS [  1.422 s]
[INFO] apiman-manager-api-war ............................. SUCCESS [  2.501 s]
[INFO] apiman-manager-api-ispn ............................ SUCCESS [  7.288 s]
[INFO] apiman-manager-api-war-wildfly8 .................... SUCCESS [  2.553 s]
[INFO] apiman-manager-ui .................................. SUCCESS [  0.005 s]
[INFO] apiman-manager-ui-war .............................. SUCCESS [01:09 min]
[INFO] apiman-manager-ui-war-wildfly8 ..................... SUCCESS [  1.108 s]
[INFO] apiman-gateway-engine-hazelcast .................... SUCCESS [ 17.843 s]
[INFO] apiman-gateway-engine-ispn ......................... SUCCESS [  1.561 s]
[INFO] apiman-gateway-engine-es ........................... SUCCESS [  1.820 s]
[INFO] apiman-gateway-engine-hawkular ..................... SUCCESS [  0.954 s]
[INFO] apiman-gateway-engine-jdbc ......................... SUCCESS [  6.078 s]
[INFO] apiman-gateway-engine-influxdb ..................... SUCCESS [  1.097 s]
[INFO] apiman-test-policies ............................... SUCCESS [  1.467 s]
[INFO] apiman-gateway-engine-policies ..................... SUCCESS [ 18.078 s]
[INFO] apiman-gateway-platforms ........................... SUCCESS [  0.005 s]
[INFO] apiman-gateway-platforms-servlet ................... SUCCESS [ 19.879 s]
[INFO] apiman-gateway-api-rest-impl ....................... SUCCESS [  1.144 s]
[INFO] apiman-gateway-platforms-war ....................... SUCCESS [  1.407 s]
[INFO] apiman-gateway-platforms-war-wildfly8 .............. SUCCESS [  0.004 s]
[INFO] apiman-gateway-platforms-war-wildfly8-api .......... SUCCESS [  0.676 s]
[INFO] apiman-gateway-platforms-war-wildfly8-gateway ...... SUCCESS [  0.667 s]
[INFO] apiman-distro-es ................................... SUCCESS [ 18.581 s]
[INFO] apiman-distro-wildfly9 ............................. SUCCESS [02:40 min]
[INFO] apiman-distro-wildfly10 ............................ SUCCESS [  4.378 s]
[INFO] apiman-distro-eap7 ................................. SUCCESS [  4.424 s]
[INFO] apiman-manager-api-war-tomcat8 ..................... SUCCESS [ 12.762 s]
[INFO] apiman-manager-ui-war-tomcat8 ...................... SUCCESS [  1.244 s]
[INFO] apiman-gateway-platforms-war-tomcat8 ............... SUCCESS [  0.003 s]
[INFO] apiman-gateway-platforms-war-tomcat8-api ........... SUCCESS [  1.708 s]
[INFO] apiman-gateway-platforms-war-tomcat8-gateway ....... SUCCESS [  0.714 s]
[INFO] apiman-distro-tomcat8 .............................. SUCCESS [  3.426 s]
[INFO] apiman-distro-wildfly8 ............................. SUCCESS [01:56 min]
[INFO] apiman-manager-api-micro ........................... SUCCESS [  6.777 s]
[INFO] apiman-manager-test ................................ SUCCESS [  0.004 s]
[INFO] apiman-manager-test-api ............................ SUCCESS [ 39.873 s]
[INFO] apiman-gateway-engine-osgi ......................... SUCCESS [  1.443 s]
[INFO] apiman-gateway-engine-prometheus ................... SUCCESS [  6.147 s]
[INFO] apiman-gateway-platforms-vertx3-common ............. SUCCESS [  1.210 s]
[INFO] apiman-gateway-engine-vertx-eb-inmemory ............ SUCCESS [  1.388 s]
[INFO] apiman-gateway-engine-vertx-polling ................ SUCCESS [  3.655 s]
[INFO] apiman-gateway-engine-vertx-shareddata ............. SUCCESS [  1.368 s]
[INFO] apiman-gateway-platforms-war-micro ................. SUCCESS [  2.126 s]
[INFO] apiman-gateway-platforms-vertx3 .................... SUCCESS [ 12.592 s]
[INFO] apiman-gateway-test ................................ SUCCESS [ 18.131 s]
[INFO] apiman-test-suite .................................. SUCCESS [  1.208 s]
[INFO] apiman-tools ....................................... SUCCESS [  0.003 s]
[INFO] apiman-tools-ddl ................................... SUCCESS [  0.840 s]
[INFO] apiman-tools-i18n .................................. SUCCESS [  1.291 s]
[INFO] apiman-tools-jdbc .................................. SUCCESS [  0.865 s]
[INFO] apiman-tools-ldap .................................. SUCCESS [  1.018 s]
[INFO] apiman-tools-services .............................. SUCCESS [  1.264 s]
[INFO] apiman-tools-server-all ............................ FAILURE [02:56 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 14:49 min
[INFO] Finished at: 2017-04-11T21:59:36+02:00
[INFO] Final Memory: 187M/429M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project apiman-tools-server-all: Could not resolve dependencies for project io.apiman:apiman-tools-server-all:pom:1.2.10-SNAPSHOT: The following artifacts could not be resolved: com.sun.faces:jsf-impl:jar:2.2.12-jbossorg-2, org.glassfish.jaxb:jaxb-core:jar:2.2.11.jbossorg-1, org.glassfish.jaxb:jaxb-runtime:jar:2.2.11.jbossorg-1, org.glassfish.jaxb:jaxb-xjc:jar:2.2.11.jbossorg-1, org.glassfish.jaxb:jaxb-jxc:jar:2.2.11.jbossorg-1, org.glassfish.jaxb:codemodel:jar:2.2.11.jbossorg-1, com.sun.xml.messaging.saaj:saaj-impl:jar:1.3.16-jbossorg-1, org.glassfish.jaxb:txw2:jar:2.2.11.jbossorg-1, com.sun.xml.bind.external:rngom:jar:2.2.11.jbossorg-1, org.slf4j:jcl-over-slf4j:jar:1.7.7.jbossorg-1, org.slf4j:slf4j-ext:jar:1.7.7.jbossorg-1, org.apache.activemq:artemis-amqp-protocol:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-cli:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-commons:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-core-client:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-dto:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-hornetq-protocol:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-hqclient-protocol:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-jms-client:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-jms-server:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-journal:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-native:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-proton-plug:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-ra:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-selector:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-server:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-service-extensions:jar:1.1.0.wildfly-011, org.apache.activemq:artemis-stomp-protocol:jar:1.1.0.wildfly-011: Could not find artifact com.sun.faces:jsf-impl:jar:2.2.12-jbossorg-2 in nexus (http://localhost:8081/nexus/content/groups/public) -> [Help 1]
[ERROR]
```

**Comment by jira-importer** [comment]
**[Marc Savy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msavy)** commented

More info:

```
[msavy@mmbp apiman](vars-section-in-config)$ mvn -version
Apache Maven 3.5.0 (ff8f5e7444045639af65f6095c62210b5713f426; 2017-04-03T20:39:06+01:00)
Maven home: /usr/local/Cellar/maven/3.5.0/libexec
Java version: 1.8.0_121, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_121.jdk/Contents/Home/jre
Default locale: en_GB, platform encoding: UTF-8
OS name: "mac os x", version: "10.12.3", arch: "x86_64", family: "mac"
```

Possibly pertinent:
- No settings.xml
- I installed the brew via Maven

If there's anything else useful I could provide, then I'm willing. I still get the same issue.

**Comment by jira-importer** [comment]
**[Marc Savy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msavy)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise) - I take it that your comment that you're not a fan of something about the build. We're always open to feedback, so if you're inclined then a ticket on our JIRA or whatever is convenient would be appreciated :-) -- https://issues.jboss.org/browse/APIMAN/. But, if you don't have the time, I understand.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Let use first try to identify the problem and find a solution...
I would suggest to first try to remove the whole $HOME/.m2/repository and try to rebuild from scratch....

**Comment by jira-importer** [comment]
**[Marc Savy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msavy)** commented

Okay, I'll give it a go and report back.

FWIW, downgrading to the previous version of Maven caused it to work
without issue on the same m2.

On Wednesday, April 12, 2017, Karl Heinz Marbaise (JIRA) <jira@apache.org>

--
Sent from mobile phone, apologies for brevity.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

By accident I hit this issue today. There are probably 2 issues in play, I was able to partly reproduce it.
The root cause is upgrade of the Modello Plugin, which contained a [fix](https://github.com/codehaus-plexus/modello/issues/5) regarding invalid XML.
What happened in my case is that one of the parents contained its content twice, i.e.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project>
...
</project>
<?xml version="1.0" encoding="UTF-8"?>
<project>
...
</project>
```

The weird thing is that the file was correct in the artifact repository, so for some reason during the download the content was duplicated.
Anyhow, I think we shouldn't use this strict setup in case of indirect poms since often we cannot change these (third party files) and Maven used to expect those kind of poms.
The other thing is that in plexus-utils we need to improve PI reading after the closing-tag.
And a completely different issue: why was the content doubled.

So the workaround is probably, clean your local repository and download the poms again.

**Comment by jira-importer** [comment]
**[Marc Savy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=msavy)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) Aha! Looks like this was the issue I was hitting. As you say, on re-download the POM was valid and the build succeeded.

It's a rather interesting issue; I guess there are no integrity checks on the POM files themselves?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

The parser used to stop with the closing project-tag, always assuming that was the end of the XML, see also MNG-6020. So yes, parsing is now more correct but exposes issues in case there is content after the closing project tag. We can only be strict like this for the pom.xml files of the reactor. Other files should be assumed that they cannot be changed.

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

Proposing that this should be considered for 3.5.1

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), is this one already addressed in master?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

If I correctly analysed that. This issue has been already fixed by the commit [0e7e3eeb597b92f1394b24e33d5933e995ee8104](https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commit;h=0e7e3eeb597b92f1394b24e33d5933e995ee8104) where modello has been upgraded to 1.9.1.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Do we need to change plexus-utils for that?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

These are 2 different issues which show the same exception. In this case, for some reason the pom file became an invalid xml file while downloading. It should have been detected when running with `-C,--strict-checksums`
However, we still need a new version of plexus for another parsing issue.

**Comment by jira-importer** [comment]
**[Brian E Fox](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brianf)** commented

I'm bumping into this on Mac with 3.5.0 and https://github.com/jeremylong/DependencyCheck/blob/master/pom.xml

**Comment by jira-importer** [comment]
**[Matthew Janssen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mttjj)** commented

I'm adding a comment to vote for a fix for this. I've been running into this non-stop on my machine after upgrading to maven 3.5.0. My coworker hasn't encountered this at all. Right now, it appears that the only way around it is to wipe the entire .m2 repo? That works for me but I don't want to have to do that every week. I was able to find some 'corrupted' poms in my repo (where the entire \<project>\</project> structure was duplicated. I deleted those poms but I still got the same failure on my project. So I'm not sure exactly what's going on. But maybe a new maven release/fix would solve my problems. Just hoping this can come sooner than later as it's really slowing down my development.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Matthew Janssen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mttjj) I have seen the issue but haven't been able to reproduce it afterwards. Your situation is interesting and could help identify the source of the issue. In general: what is the difference between you and your co-worker? Different IDE? Running Maven only from IDE or only from commandline or mixed? Build with multiple threads? Different OS? Different Java version?

**Comment by jira-importer** [comment]
**[Matthew Janssen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mttjj)** commented

Thanks for the reply [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte). That's what's so baffling about this to me. I made sure that my system is set up just like his.
Here is the result of my co-worker's `mvn -version` run: https://gist.github.com/mttjj/54ec4da3e56358aff1c7e32278a0c23b
Here is the result of my `mvn -version` run: https://gist.github.com/mttjj/b69618a2d9fd6cfd78413da86287bbfc
The only differences I'm seeing are the installation directories and the java version.

As far as your other questions: if I recall correctly my co-worker runs commands both from within Eclipse (he's using Mars) and the command line (but I think it's from the command line a majority of the time). I, personally, run maven commands exclusively from the command line. We don't have any special thread configuration so it's just whatever the maven default is which I assume is single-threaded. And the OS/Java versions are included in the gists above.

Some other notes: the `settings.xml` files in our `maven\conf` folder are just the standard settings (no changes for either of us) and neither of us have a `settings.xml` in our m2 repo either.
Both of us started from clean m2 repos right after we updated to maven 3.5.0. We didn't run maven installs on all of the exact same jars in the exact same order but the project we're working on now has us working with the same general jars so we had installed similar projects over the same time period.

For me, this error doesn't occur on every project I try to install, it just crops up sometime and then I can never install that project without clearing my m2 repo. I've wiped my m2 repo multiple times. After each time, there's a high probability that _will_ be able to install the troublesome jar but a different jar will inevitably fail at some point.

Let me know if there's any more info I can provide to help get to the bottom of this!

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Matthew Janssen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mttjj), just to be sure: you're both using default `settings.xml`, so there's no repository-manager in the middle, nor a proxy. You're both directly connected to Maven Central, right?
It still looks like a nasty race condition to me when downloading files in parallel.

It is unlikely that the different install location is the issue, the location of the local repository would be more likely to be an issue (local or part of roaming profile).
The Java version could be an issue, we've seen this before, although I don't expect it to be. Could be interesting to try to sync these two, just to remove another variable.

**Comment by jira-importer** [comment]
**[Matthew Janssen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mttjj)** commented

To answer your first questions: yes to all. Technically, our company actually mirrors Maven Central so when we are hitting maven central we are actually hitting our own version of it but it's effectively the same thing.

Interestingly, I haven't encountered the issue since the last time I posted here. I know I just jinxed myself by saying that but we'll see. I still think there's a problem here but I haven't seen it in a while!

Thanks for checking in.

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

[Matthew Janssen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mttjj)
What mirror software are you using? There are some mirrors available that have a "feature" to rewrite the pom.xml on the fly and remove and inject things in the poms, e.g. removing the \<repositories> section or even injecting a replacement one.

IIRC neither nexus nor archivia have that "feature" (or if they have the feature they do not have it enabled by default) but if you are using the other one, it may be a useful data-point

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc) I've seen it once when I was connected to a Nexus Instance. The `pom.xml` was correct in the repository manager, but "doubled" in my local repository.

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) were you running parallel builds or two maven sessions concurrently? (Perhaps your IDE was downloading POMs concurrently with your maven CLI build?)

The rebased https://builds.apache.org/blue/organizations/jenkins/maven-3.x-jenkinsfile/detail/MNG-6216/5/pipeline which has the plexus update should handle the doubled pom's (though it would be good to find out where they were originating)

I am now wondering if there could be some IDE background downloading and not doing in a concurrent safe way

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

I've merged https://git-wip-us.apache.org/repos/asf?p=maven.git;a=commitdiff;h=c9a288d8b1090fa957d6caccc12f0bf13bb5e267 given that the build passed

**Comment by jira-importer** [comment]
**[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) we need to decide whether to resolve this issue or not. I propose we resolve this issue and spin off a task to try and identify the root cause of the corrupted POMs

**Comment by jira-importer** [comment]
**[Matthew Janssen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mttjj)** commented

[Stephen Connolly](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stephenc) As far as I know, our company uses Nexus as our repository. But I don't know a whole lot about it since that's a completely separate org that handles our build factories.

And +1 on resolving the issue in the short term and identifying the problem over the long term.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

> were you running parallel builds or two maven sessions concurrently? (Perhaps your IDE was downloading POMs concurrently with your maven CLI build?)
> I am now wondering if there could be some IDE background downloading and not doing in a concurrent safe way
> IDE (probably Eclipse at that time) open and running Maven via commandline. So yes, there's a chance that 2 processes are trying to do the same thing, but they are somehow appended at the end.

> The rebased https://builds.apache.org/blue/organizations/jenkins/maven-3.x-jenkinsfile/detail/MNG-6216/5/pipeline which has the plexus update should handle the doubled pom's (though it would be good to find out where they were originating)
> That's the wrong conclusion. It contains a fix for the AIOOBE when there's a XML process instruction in a large pom file.

So we have the same exception, but with different causes.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Can someone verify this with 3.6.2-SNAPSHOT? Many improvements landed in Plexus Utils recently.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Does this still occur? If this is a concurrency issue it can/well be solved with MNG-2802.

**Comment by jira-importer** [comment]
**[Matthew Janssen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mttjj)** commented

I, personally, have not encountered this issue since my posts - geez - 3 years ago. I have used all subsequent maven releases (currently using 3.6.3) and this issue has not resurfaced. I've also gotten a new computer and wiped my .m2 repo many times since my last post. Whatever the issue was, I'm not seeing it anymore.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), are we OK to close it?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

There are 2 rootcauses for getting the ArrayIndexOutOfBoundsException and AFAIK MNG-6281 is still an issue, unless something has been changed in Maven Resolver, assuming the real issue is there.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

The only thing I can think of is concurrency (race conditions) issues in Resolver. Version 1.6.0 will provide two concurrency safe contexts to avoid such issues.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

It is indeed a race condition issue, but IIRC with a simple project I was able to hit it somewhere between 20% to 33% of the times. That might be often enough for an IT.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I'd be happy to analyze if you could provide me the sample project.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I believe that the actual race condition has been fixed by MNG-2802 with two sync contexts available.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Maven Resolver 1.6.2 will be available with Maven 3.8.2 so this has been addressed in 3.8.x too.
